server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsFinder.java

```diff
@@ -87,16 +87,10 @@ public SettingsFinder(DbClient dbClient, PropertyDefinitions definitions) {
     properties.stream()
       .filter(propertyDto -> definitions.get(propertyDto.getKey()) != null)
       .filter(propertyDto -> definitions.get(propertyDto.getKey()).type().equals(PROPERTY_SET))
-      .forEach(propertyDto -> definitions.get(propertyDto.getKey()).fields()
-        .forEach(field -> COMMA_SPLITTER.splitToList(propertyDto.getValue())
-          .forEach(setId -> propertySetKeys.add(generatePropertySetKey(propertyDto.getKey(), setId, field.key())))));
+      .forEach(propertyDto -> propertySetKeys.addAll(extractPropertySetKeys(propertyDto, definitions.get(propertyDto.getKey()))));
```

Comment:
what about a `flatMap(property -> extractPropertySetKey(property, definitions.get(property.getKey())))` and `collect(Collectors.toSet())` in the end?